File: resources/ios/SelectRenderer.swift
import SwiftUI

enum SelectRendererEvent: Equatable {
    case press(callbackId: Int, nodeId: Int)

    var wireName: String { "PRESS" }
}

struct SelectRendererEvents {
    let send: @MainActor (SelectRendererEvent) -> Void

    static let native = SelectRendererEvents { event in
        switch event {
        case let .press(callbackId, nodeId):
            NativeUIBridge.sendPressEvent(callbackId, nodeId: nodeId)
        }
    }
}

struct SelectRendererConfiguration: Equatable {
    let nodeID: Int
    let optionValues: [String]
    let optionLabels: [String]
    let optionEnabled: [Bool]
    let optionCallbacks: [Int]
    let selectedValues: [String]
    let valueType: String
    let searchEnabled: Bool
    let disabled: Bool
    let label: String
    let placeholder: String
    let helper: String
    let error: String
    let required: Bool
    let accessibilityLabel: String
    let accessibilityHint: String

    init(node: NativeUINode) {
        let props = node.props
        let values = props.getStringList("option_values")

        nodeID = node.id
        optionValues = values
        optionLabels = props.getStringList("option_labels")
        optionEnabled = props.getStringList("option_enabled").map { $0 == "1" }
        optionCallbacks = props.getStringList("option_callbacks").map { Int($0) ?? 0 }
        selectedValues = props.getStringList("selected_values")
        valueType = props.getString("value_type")
        searchEnabled = props.getBool("search_enabled")
        disabled = props.getBool("disabled") || values.isEmpty
        label = props.getString("label")
        placeholder = props.getString("placeholder")
        helper = props.getString("helper")
        error = props.getString("error")
        required = props.getBool("required")
        let explicitAccessibilityLabel = props.getString("a11y_label")
        accessibilityLabel = explicitAccessibilityLabel.isEmpty
            ? props.getString("label")
            : explicitAccessibilityLabel
        accessibilityHint = props.getString("a11y_hint")
    }
}

struct SelectRendererState: Equatable {
    var configuration: SelectRendererConfiguration
    private(set) var isAwaitingPublication = false

    init(node: NativeUINode) {
        configuration = SelectRendererConfiguration(node: node)
    }

    mutating func userSelected(_ index: Int) -> SelectRendererEvent? {
        guard !isAwaitingPublication,
              !configuration.disabled,
              configuration.optionValues.indices.contains(index),
              configuration.optionValues[index] != configuration.selectedValues.first,
              configuration.optionEnabled.indices.contains(index),
              configuration.optionEnabled[index],
              configuration.optionCallbacks.indices.contains(index),
              configuration.optionCallbacks[index] != 0
        else {
            return nil
        }

        isAwaitingPublication = true

        return .press(
            callbackId: configuration.optionCallbacks[index],
            nodeId: configuration.nodeID
        )
    }

    @discardableResult
    mutating func serverPublished(_ tree: NativeUITree) -> Bool {
        guard let node = Self.findNode(id: configuration.nodeID, in: tree.root) else {
            return false
        }

        let previousSelection = configuration.selectedValues
        configuration = SelectRendererConfiguration(node: node)
        isAwaitingPublication = false

        return previousSelection != configuration.selectedValues
    }

    private static func findNode(id: Int, in node: NativeUINode) -> NativeUINode? {
        if node.id == id { return node }

        for child in node.children {
            if let match = findNode(id: id, in: child) { return match }
        }

        return nil
    }
}

struct SelectRenderer: View {
    @ObservedObject private var themeStore = NativeUITheme.shared
    @ObservedObject private var bridge = NativeUIBridge.shared
    @Environment(\.colorScheme) private var colorScheme
    @State private var rendererState: SelectRendererState

    private let events: SelectRendererEvents

    init(node: NativeUINode, events: SelectRendererEvents = .native) {
        _rendererState = State(initialValue: SelectRendererState(node: node))
        self.events = events
    }

    var body: some View {
        let configuration = rendererState.configuration
        let theme = themeStore.resolve(for: colorScheme)

        FirstlightSelectField(
            labels: configuration.optionLabels,
            values: configuration.optionValues,
            optionEnabled: configuration.optionEnabled,
            selectedValues: configuration.selectedValues,
            searchEnabled: configuration.searchEnabled,
            disabled: configuration.disabled,
            awaitingPublication: rendererState.isAwaitingPublication,
            label: configuration.label,
            placeholder: configuration.placeholder,
            helper: configuration.helper,
            error: configuration.error,
            required: configuration.required,
            tokens: FirstlightSelectTokens.from(theme: theme),
            accessibilityLabel: configuration.accessibilityLabel,
            accessibilityHint: configuration.accessibilityHint,
            onSelection: { index in
                guard let event = rendererState.userSelected(index) else { return }
                events.send(event)
            }
        )
        .onReceive(bridge.$currentTree.compactMap { $0 }) { tree in
            rendererState.serverPublished(tree)
        }
    }
}

File: resources/android/SelectRenderer.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import com.nativephp.mobile.ui.nativerender.NativeUIBridge
import com.nativephp.mobile.ui.nativerender.NativeUINode
import com.nativephp.mobile.ui.nativerender.NativeUITree
import com.nativephp.plugins.native_ui.NativeUITheme

sealed interface SelectRendererEvent {
    val wireName: String

    data class Press(
        val callbackId: Int,
        val nodeId: Int,
    ) : SelectRendererEvent {
        override val wireName = "PRESS"
    }
}

class SelectRendererEvents(
    private val send: (SelectRendererEvent) -> Unit,
) {
    fun dispatch(event: SelectRendererEvent) = send(event)

    companion object {
        val native = SelectRendererEvents { event ->
            when (event) {
                is SelectRendererEvent.Press -> NativeUIBridge.sendPressEvent(
                    event.callbackId,
                    event.nodeId,
                )
            }
        }
    }
}

data class SelectRendererConfiguration(
    val nodeId: Int,
    val optionValues: List<String>,
    val optionLabels: List<String>,
    val optionEnabled: List<Boolean>,
    val optionCallbacks: List<Int>,
    val selectedValues: List<String>,
    val valueType: String,
    val searchEnabled: Boolean,
    val disabled: Boolean,
    val label: String,
    val placeholder: String,
    val helper: String,
    val error: String,
    val required: Boolean,
    val accessibilityLabel: String,
    val accessibilityHint: String,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        optionValues = node.props.getStringList("option_values"),
        optionLabels = node.props.getStringList("option_labels"),
        optionEnabled = node.props.getStringList("option_enabled").map { it == "1" },
        optionCallbacks = node.props.getStringList("option_callbacks").map { it.toIntOrNull() ?: 0 },
        selectedValues = node.props.getStringList("selected_values"),
        valueType = node.props.getString("value_type"),
        searchEnabled = node.props.getBool("search_enabled"),
        disabled = node.props.getBool("disabled") || node.props.getStringList("option_values").isEmpty(),
        label = node.props.getString("label"),
        placeholder = node.props.getString("placeholder"),
        helper = node.props.getString("helper"),
        error = node.props.getString("error"),
        required = node.props.getBool("required"),
        accessibilityLabel = node.props.getString("a11y_label").ifEmpty {
            node.props.getString("label")
        },
        accessibilityHint = node.props.getString("a11y_hint"),
    )
}

class SelectRendererState(node: NativeUINode) {
    var configuration by mutableStateOf(SelectRendererConfiguration(node))
        private set

    var isAwaitingPublication by mutableStateOf(false)
        private set

    fun userSelected(index: Int): SelectRendererEvent? {
        val selectedValue = configuration.selectedValues.firstOrNull()

        if (isAwaitingPublication ||
            configuration.disabled ||
            !configuration.optionValues.indices.contains(index) ||
            !configuration.optionEnabled.getOrElse(index) { false } ||
            configuration.optionValues[index] == selectedValue
        ) {
            return null
        }

        val callback = configuration.optionCallbacks.getOrNull(index)?.takeIf { it != 0 }
            ?: return null

        isAwaitingPublication = true
        return SelectRendererEvent.Press(callback, configuration.nodeId)
    }

    fun serverPublished(tree: NativeUITree): Boolean {
        val node = tree.root.findSelectNode(configuration.nodeId) ?: return false
        val previousSelection = configuration.selectedValues
        configuration = SelectRendererConfiguration(node)
        isAwaitingPublication = false
        return previousSelection != configuration.selectedValues
    }
}

object FirstlightSelectRenderer {
    @Composable
    fun Render(
        node: NativeUINode,
        modifier: Modifier,
        events: SelectRendererEvents = SelectRendererEvents.native,
    ) {
        val rendererState = remember(node.id) { SelectRendererState(node) }
        val publishedTree = NativeUIBridge.currentTree.value

        LaunchedEffect(publishedTree) {
            publishedTree?.let(rendererState::serverPublished)
        }

        val tokens = if (isSystemInDarkTheme()) NativeUITheme.dark else NativeUITheme.light

        FirstlightSelectControl(
            configuration = rendererState.configuration,
            awaitingPublication = rendererState.isAwaitingPublication,
            onSelection = { index ->
                rendererState.userSelected(index)?.let(events::dispatch)
            },
            modifier = modifier,
            tokens = tokens,
        )
    }
}

private fun NativeUINode.findSelectNode(id: Int): NativeUINode? {
    if (this.id == id) return this
    children.forEach { child -> child.findSelectNode(id)?.let { return it } }
    return null
}
